torchvision/prototype/transforms/_geometry.py

```diff
@@ -448,43 +449,62 @@ def __init__(
+        input_height, input_width = height, width
+
+        if self.pad_if_needed:
+            # pad width if needed
+            if width < output_width:
+                width += 2 * (output_width - width)
+            # pad height if needed
+            if height < output_height:
+                height += 2 * (output_height - height)
 
         if height + 1 < output_height or width + 1 < output_width:
             raise ValueError(
                 f"Required crop size {(output_height, output_width)} is larger then input image size {(height, width)}"
             )
 
         if width == output_width and height == output_height:
-            return dict(top=0, left=0, height=height, width=width)
+            return dict(top=0, left=0, height=height, width=width, input_width=input_width, input_height=input_height)
 
         top = torch.randint(0, height - output_height + 1, size=(1,)).item()
         left = torch.randint(0, width - output_width + 1, size=(1,)).item()
-        return dict(top=top, left=left, height=output_height, width=output_width)
-
-    def _transform(self, inpt: Any, params: Dict[str, Any]) -> Any:
-        return F.crop(inpt, **params)
 
-    def forward(self, *inputs: Any) -> Any:
-        sample = inputs if len(inputs) > 1 else inputs[0]
+        return dict(
+            top=top,
+            left=left,
+            height=output_height,
+            width=output_width,
+            input_width=input_width,
+            input_height=input_height,
+        )
 
+    def _transform(self, inpt: Any, params: Dict[str, Any]) -> Any:
         if self.padding is not None:
-            sample = F.pad(sample, padding=self.padding, fill=self.fill, padding_mode=self.padding_mode)
-
-        image = query_image(sample)
-        _, height, width = get_image_dimensions(image)
+            inpt = F.pad(inpt, padding=self.padding, fill=self.fill, padding_mode=self.padding_mode)
 
         if self.pad_if_needed:
-            # pad the width if needed
-            if width < self.size[1]:
-                padding = [self.size[1] - width, 0]
-                sample = F.pad(sample, padding=padding, fill=self.fill, padding_mode=self.padding_mode)
-            # pad the height if needed
-            if height < self.size[0]:
-                padding = [0, self.size[0] - height]
-                sample = F.pad(sample, padding=padding, fill=self.fill, padding_mode=self.padding_mode)
-
-        return super().forward(sample)
+            input_width, input_height = params["input_width"], params["input_height"]
+            if input_width < self.size[1]:
+                padding = [self.size[1] - input_width, 0]
+                inpt = F.pad(inpt, padding=padding, fill=self.fill, padding_mode=self.padding_mode)
+            if input_height < self.size[0]:
+                padding = [0, self.size[0] - input_height]
+                inpt = F.pad(inpt, padding=padding, fill=self.fill, padding_mode=self.padding_mode)
```

Comment:
Isn't it possible to combine some of these 3 pads together? I'm afraid that doing 3 separate ones here will end up being slow.

Reply:
Yes, maybe it could be possible. Without correctness tests do you think we should try to do that right now ?

Reply:
You are right. Let's just add a TODO to optimize. This is going to give us a nice speed improvement over the existing API, as we use this practice on the old API. We should be on the look out for such patterns that we can optimize and mark them with performance TODOs for the next milestone.